include/psa/crypto_struct.h

```diff
@@ -245,20 +261,40 @@ typedef struct psa_tls12_prf_key_derivation_s
 #endif /* MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF) ||
         * MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS */
 
+typedef struct
+{
+    union
+    {
+        /* Make the union non-empty even with no supported algorithms. */
+        uint8_t MBEDTLS_PRIVATE(dummy);
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF) || \
+    defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS)
+        psa_tls12_prf_key_derivation_inputs_t MBEDTLS_PRIVATE(prf);
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF ||
+          MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS */
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
+        psa_hkdf_key_derivation_inputs_t MBEDTLS_PRIVATE(hkdf);
+    };
+#endif /* MBEDTLS_PSA_BUILTIN_ALG_HKDF */
+} psa_crypto_driver_key_derivation_inputs_t;
+
 struct psa_key_derivation_s
 {
     psa_algorithm_t MBEDTLS_PRIVATE(alg);
     unsigned int MBEDTLS_PRIVATE(can_output_key) : 1;
     size_t MBEDTLS_PRIVATE(capacity);
+
     union
```

Comment:
> The context is now organized by stage first, and by implementation second, instead of being organized by algorithm first  and more or less by stage inside each algorithm

Currently (`development`) the `context` is organized by algorithm(hkdf vs prf) first:

```
    union
    {
        /* Make the union non-empty even with no supported algorithms. */
        uint8_t MBEDTLS_PRIVATE(dummy);
#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
        psa_hkdf_key_derivation_t MBEDTLS_PRIVATE(hkdf);
#endif
#if defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF) || \
    defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS)
        psa_tls12_prf_key_derivation_t MBEDTLS_PRIVATE(tls12_prf);
#endif
    } MBEDTLS_PRIVATE(ctx);
```

What do you exactly mean by: `stage`, `implementation`, `algorithm`?

Reply:
Stage: 1 = gathering inputs, 2 = producing output

Algorithm: HKDF, TLS12_PRF, ...

Implementation: currently one for each algorithm, but the next step after this task is to support drivers